### PR TITLE
expression: refactor cache logic for `builtinRegexpSubstrFuncSig`

### LINKED ISSUE #49583: refactor cache implement for function `regexp_substr`
## Enhancement

see #49464

`regexp_substr` caches some intermediate `regexp.Regexp` inside the function when some args are constant. Some issues for the current implement:

1. The cache logic is coupled with plan cahe, and it is hard to say the function is `stateless`.
2. When plan cahe is enabled, we cannot cache the object `regexp.Regexp` some times. For example, when the regex pattern is from the variable.

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #49583

Problem Summary:

See #49583 and #49464

### What changed and how does it work?

1. Add `CtxID` to `StatementContext` to return a unique ID for statement context. 
2. Introduce `builtinFuncCache` to provide a cache mechanism for built function. It uses context ID to separate cache between different statements so that we can still cache some immediate states even if plan cache is enabled now.
3. Introduce `regexpNewBaseFuncSig` as a new base for regex related functions. We'll use `regexpNewBaseFuncSig` to replace the old `regexpBaseFuncSig` in the next several PRs.
4. refactor `builtinRegexpSubstrFuncSig` to use `regexpNewBaseFuncSig` to manage the cache, so that:
  - It's implement can decouple with plan cache
  - It can still cache the parsed `*regexp.Regexp` when plan cache is on and regex pattern is read from a variable

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

We did a benchmark test and the performance of `regexp_substr` increased:

```
func BenchmarkRegexpSubStr(b *testing.B) {
	store := testkit.CreateMockStore(b)
	tk := testkit.NewTestKit(b, store)
	tk.MustExec("use test")
	tk.MustExec("create table t(a varchar(256))")
	for i := 0; i < b.N; i++ {
		tk.MustExec(fmt.Sprintf("insert into t values('pingcaper%d@pingcap.com')", i))
	}
	tk.MustExec(`set @regexp='^\\w+([-+.]\\w+)*@\\w+([-.]\\w+)*\\.\\w+([-.]\\w+)*$'`)
	tk.MustExec("prepare s from 'select regexp_substr(a, ?) from t'")

	b.ResetTimer()
	tk.MustQuery("execute s using @regexp")
	b.StopTimer()
}
```

Master:

```
go test -bench='^BenchmarkRegexpSubStr$' -run='^$' -benchtime=1000000x --tags=intest
...
 1000000	      1804 ns/op
PASS
ok  	github.com/pingcap/tidb/pkg/expression/integration_test	64.611s
```

This PR:
```
go test -bench='^BenchmarkRegexpSubStr$' -run='^$' -benchtime=1000000x --tags=intest
...
 1000000	       333.3 ns/op
PASS
ok  	github.com/pingcap/tidb/pkg/expression/integration_test	64.772s
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
improve the performance of function `regexp_substr` in some cases when plan cache enabled
```

### CODE REVIEW

File: pkg/expression/builtin.go [hunk 1]
```diff
@@ -1017,3 +1019,46 @@ func (b *baseBuiltinFunc) MemoryUsage() (sum int64) {
...
+	if p := c.cached.Load(); p != nil && p.ctxID == ctxID {
+		return p.item, true
+	}
+	return v, false
+}
+
+func (c *builtinFuncCache[T]) getOrInitCache(ctx EvalContext, constructCache func() (T, error)) (T, error) {
+	intest.Assert(constructCache != nil)
+	ctxID := ctx.GetSessionVars().StmtCtx.CtxID()
+	if item, ok := c.getCache(ctxID); ok {
```
Comment (xzhangxian1008) [inline pkg/expression/builtin.go:1043]: Why we double check the cache outside of the lock protection?

File: pkg/sessionctx/stmtctx/stmtctx.go [hunk 2]
```diff
@@ -425,26 +429,36 @@ type StatementContext struct {
 
 // NewStmtCtx creates a new statement context
 func NewStmtCtx() *StatementContext {
```
Comment (YangKeao) [inline pkg/sessionctx/stmtctx/stmtctx.go:431]: Not related with this PR. Is it possible to use `NewStmtCtxWithTimeZone(time.UTC)` to implement `NewStmtCtx`?

**Review by YangKeao — APPROVED** [review]
LGTM

File: pkg/expression/builtin_test.go [hunk 3]
```diff
@@ -161,6 +163,71 @@ func TestDisplayName(t *testing.T) {
 	require.Equal(t, "other_unknown_func", GetDisplayName("other_unknown_func"))
 }
 
+func TestBuiltinFuncCache(t *testing.T) {
```
Comment (xzhangxian1008) [inline pkg/expression/builtin_test.go:166]: I think it's better to add some concurrent tests.

File: pkg/expression/builtin.go (hunk 1 above)
Comment (lcwangchao) [reply pkg/expression/builtin.go:1043]: To achieve a better performance, ensure a concurrent call of `getOrInitCache` will not block each other.

File: pkg/sessionctx/stmtctx/stmtctx.go (hunk 2 above)
Comment (lcwangchao) [reply pkg/sessionctx/stmtctx/stmtctx.go:431]: Yes, I think it is a better way when we have more fields to init

File: pkg/expression/builtin.go [hunk 4]
```diff
@@ -1017,3 +1019,46 @@ func (b *baseBuiltinFunc) MemoryUsage() (sum int64) {
...
+		return item, nil
+	}
+
+	c.Lock()
```
Comment (YangKeao) [inline pkg/expression/builtin.go:1047]: Maybe we can implement it in CAS pattern as it's already an atomic pointer.

```go
	old := c.cached.Load()
	if old.ctxID == ctxID {
		return old.item, nil
	}

	newItem, err := constructCache()
	if err != nil {
		var def T
		return def, err
	}
	newCache := &builtinFuncCacheItem[T]{
		item:  newItem,
		ctxID: ctxID,
	}
	for !c.cached.CompareAndSwap(old, newCache) {
		old := c.cached.Load()
		if old.ctxID == ctxID {
			return old.item, nil
		}
	}

	return newItem, nil
```

But actually, I have no opinion which one is better :thinking:.

Comment (lcwangchao) [reply pkg/expression/builtin.go:1047]: The `constructCache` seems to have a probability of being called more than once even if ctxID not change (but I think it is also acceptable).

File: pkg/expression/builtin_test.go (hunk 3 above)
Comment (lcwangchao) [reply pkg/expression/builtin_test.go:196]: A new test case `TestBuiltinFuncCacheConcurrency` added

**Review by xzhangxian1008 — APPROVED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**